### PR TITLE
fix(datasource/go): consider all available datasources for go-import

### LINKED ISSUE #16543: Go modules: custom import path behind proxy, <meta name="go-import" content=""> not handled correctly
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.111.0

### Please select which platform you are using if self-hosting.

gitlab.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

15.2.0-pre e3eee12ca75

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

I'm readressing #12394.
since most servers now removed the deprecated `go-source` meta tag, it is now important to consider all available datasources for the `go-import` meta tag as well:
```html
<meta name="go-import" content="example.org git https://code.org/r/p/exproj">
```
See https://go.dev/ref/mod#serving-from-proxy for official documentation.

For the old `go-source` tag, renovate already supports all datasources, but for `go-import`, the datasource is hardcoded on `GithubTagsDatasource`.

Modules that are hosted elsewhere can not be checked by renovate so far.

### Relevant debug logs

<details><summary>Logs</summary>

```
{"name":"renovate","hostname":"runner-648zxhn3-project-37675849-concurrent-0","pid":12,"level":20,"logContext":"ITM7nJM1W5svQG1F00Sj0","repository":"some/private/repository","goModule":"go.myserver.com/some/module","goImportURL":"https://gitlab.com/some/private/module.git","msg":"Go lookup import url","time":"2022-07-11T12:59:10.494Z","v":0}
{"name":"renovate","hostname":"runner-648zxhn3-project-37675849-concurrent-0","pid":12,"level":20,"logContext":"ITM7nJM1W5svQG1F00Sj0","repository":"some/private/repository","datasource":"go","packageName":"go.myserver.com/some/module","err":{"message":"Field 'repository' doesn't exist on type 'Query'","stack":"Error: Field 'repository' doesn't exist on type 'Query'\n    at CacheableGithubTags.query (/usr/src/app/node_modules/renovate/lib/modules/datasource/github-releases/cache/cache-base.ts:178:22)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at CacheableGithubTags.getItems (/usr/src/app/node_modules/renovate/lib/modules/datasource/github-releases/cache/cache-base.ts:267:23)\n    at GithubTagsDatasource.getReleases (/usr/src/app/node_modules/renovate/lib/modules/datasource/github-tags/index.ts:84:25)\n    at GoDirectDatasource.getReleases (/usr/src/app/node_modules/renovate/lib/modules/datasource/go/releases-direct.ts:59:15)\n    at /usr/src/app/node_modules/renovate/lib/util/cache/package/decorator.ts:78:20\n    at /usr/src/app/node_modules/renovate/lib/util/cache/package/decorator.ts:78:20\n    at fetchReleases (/usr/src/app/node_modules/renovate/lib/modules/datasource/index.ts:300:13)\n    at getPkgReleases (/usr/src/app/node_modules/renovate/lib/modules/datasource/index.ts:352:7)\n    at lookupUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/lookup/index.ts:80:32)\n    at fetchDepUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/fetch.ts:58:15)\n    at /usr/src/app/node_modules/p-map/index.js:57:22"},"msg":"Datasource unknown error","time":"2022-07-11T12:59:10.745Z","v":0}
{"name":"renovate","hostname":"runner-648zxhn3-project-37675849-concurrent-0","pid":12,"level":20,"logContext":"ITM7nJM1W5svQG1F00Sj0","repository":"some/private/repository","dependency":"go.myserver.com/some/module","packageFile":"go.mod","msg":"Failed to look up dependency go.myserver.com/some/module","time":"2022-07-11T12:59:10.745Z","v":0}
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

### PR DESCRIPTION
**Pull request by tessig**
## Changes

For go modules behind a proxy, the source is given via meta tag `go-import`. For the moment, only Github is supported as datasource.
This uses the datasource selection already in place for the old `go-source` tag and falls back to the hardcoded github, if no datasource was found.

## Context

Closes #16543

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Hi, thanks for the PR. What's the next step here? I'm a little confused as I was expecting to see `go-import` and/or `go-source` strings within the changes

**Comment by tessig** [comment]
Hi, the checks for `go-source` and `go-import` are already there (lines 75 and 162). I just moved the datasource detection from `goSourceHeader` func to a separate function and reused it in `goImportHeader`.

Next steps are:
* add a test case to `base.spec.ts`
* set up a public example

**Comment by rarkins** [comment]
Is the test repository meant to show an *error*? I.e. the non working case and not the working case?

**Comment by tessig** [comment]
Yes, the test repo shows the current state. Was unsure if you wanted to see both results.
But I just ran this branch against the test repository.

**Comment by rarkins** [comment]
Are there any public go modules/packages this can be tested against?

**Comment by rarkins** [comment]
Yes but if you wrote both the code and the test then it's possible both are wrong. It would be entirely possible that your new code works with no module in the world except the one you wrote specifically to test against. Is there really no publicly available module (a "real" package) which benefits from this fix?

File: lib/modules/datasource/go/base.ts [hunk 1]
```diff
@@ -173,18 +180,26 @@ export class BaseGoDatasource {
...
-      datasource: GithubTagsDatasource.id,
-      registryUrl: `${parsedUrl.protocol}//${parsedUrl.host}`,
-      packageName,
-    };
+    const datasource = this.detectDatasource(
+      goImportURL.replace(regEx(/\.git$/), ''),
+      goModule
+    );
+    // fall back to old behaviour if detection did not work
+    if (datasource === null) {
```
Comment (rarkins) [inline lib/modules/datasource/go/base.ts:188]:
```suggestion
    if (datasource !== null) {
      return datasource;
    }
```

Return early to reduce white space

**Review by rarkins — CHANGES_REQUESTED** [review]

**Review by rarkins — APPROVED** [review]

**Review by zharinov — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**